File: backend/app/trace_store.py
import json
import sqlite3
import time
from pathlib import Path
from typing import Any, Dict, List, Optional


ROOT = Path(__file__).resolve().parents[2]
DB_PATH = ROOT / "data" / "traces.sqlite3"


class TraceStore:
    def __init__(self, path: Path = DB_PATH):
        self.path = path
        self.path.parent.mkdir(parents=True, exist_ok=True)
        self._init()

    def _connect(self):
        return sqlite3.connect(self.path)

    def _init(self) -> None:
        with self._connect() as db:
            db.execute(
                """
                create table if not exists traces (
                  trace_id text primary key,
                  started_at real not null,
                  ended_at real,
                  client_protocol text,
                  client_name text,
                  incoming_model text,
                  resolved_profile_id text,
                  resolved_role text,
                  upstream_provider_id text,
                  upstream_model text,
                  status text,
                  latency_ms integer,
                  usage_json text,
                  sanitizer_json text,
                  steps_json text,
                  request_json text,
                  response_json text
                )
                """
            )

    def put(self, record: Dict[str, Any]) -> None:
        record = dict(record)
        record.setdefault("started_at", time.time())
        with self._connect() as db:
            db.execute(
                """
                insert or replace into traces values
                (:trace_id, :started_at, :ended_at, :client_protocol, :client_name,
                 :incoming_model, :resolved_profile_id, :resolved_role,
                 :upstream_provider_id, :upstream_model, :status, :latency_ms,
                 :usage_json, :sanitizer_json, :steps_json, :request_json, :response_json)
                """,
                self._row(record),
            )

    def list(self, limit: int = 100) -> List[dict]:
        with self._connect() as db:
            db.row_factory = sqlite3.Row
            rows = db.execute(
                "select * from traces order by started_at desc limit ?", (limit,)
            ).fetchall()
        return [self._decode(dict(row)) for row in rows]

    def get(self, trace_id: str) -> Optional[dict]:
        with self._connect() as db:
            db.row_factory = sqlite3.Row
            row = db.execute("select * from traces where trace_id = ?", (trace_id,)).fetchone()
        return self._decode(dict(row)) if row else None

    def clear(self) -> None:
        with self._connect() as db:
            db.execute("delete from traces")

    def _row(self, record: Dict[str, Any]) -> dict:
        def dumps(key):
            return json.dumps(record.get(key) or {}, ensure_ascii=False)

        return {
            "trace_id": record["trace_id"],
            "started_at": record.get("started_at"),
            "ended_at": record.get("ended_at"),
            "client_protocol": record.get("client_protocol"),
            "client_name": record.get("client_name"),
            "incoming_model": record.get("incoming_model"),
            "resolved_profile_id": record.get("resolved_profile_id"),
            "resolved_role": record.get("resolved_role"),
            "upstream_provider_id": record.get("upstream_provider_id"),
            "upstream_model": record.get("upstream_model"),
            "status": record.get("status"),
            "latency_ms": record.get("latency_ms"),
            "usage_json": dumps("usage"),
            "sanitizer_json": dumps("sanitizer"),
            "steps_json": dumps("steps"),
            "request_json": dumps("request"),
            "response_json": dumps("response"),
        }

    def _decode(self, row: dict) -> dict:
        for src, dst in [
            ("usage_json", "usage"),
            ("sanitizer_json", "sanitizer"),
            ("steps_json", "steps"),
            ("request_json", "request"),
            ("response_json", "response"),
        ]:
            row[dst] = json.loads(row.pop(src) or "{}")
        return row

File: worker/src/storage/traceStore.ts
import type { TraceRecord } from "../types/trace";
import { redact } from "../utils/redact";
import { parseJsonColumn } from "./d1";

interface TraceRow {
  request_id: string;
  method: string;
  path: string;
  alias?: string | null;
  target_model?: string | null;
  provider_id?: string | number | null;
  status: number;
  latency_ms: number;
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
  error?: string | null;
  request_json?: string | null;
  response_json?: string | null;
  steps_json?: string | null;
  created_at: string;
}

export class TraceStore {
  constructor(private readonly db: D1Database) {}

  async create(record: TraceRecord & { method?: string; path?: string }): Promise<void> {
    const safe = redact(record);
    await this.db
      .prepare(
        `INSERT INTO traces
          (request_id, alias, target_model, provider_id, method, path, status, latency_ms,
           prompt_tokens, completion_tokens, total_tokens, error, request_json, response_json, steps_json)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      )
      .bind(
        safe.trace_id,
        safe.incoming_model ?? null,
        safe.upstream_model ?? null,
        safe.upstream_provider_id ?? null,
        safe.method ?? "POST",
        safe.path ?? "/",
        safe.status === "success" ? 200 : 500,
        safe.latency_ms ?? 0,
        safe.usage?.inputTokens ?? 0,
        safe.usage?.outputTokens ?? 0,
        safe.usage?.totalTokens ?? 0,
        safe.status === "error" ? JSON.stringify(safe.response ?? safe.request ?? {}) : null,
        JSON.stringify(safe.request ?? {}),
        JSON.stringify(safe.response ?? {}),
        JSON.stringify(safe.steps ?? []),
      )
      .run();
  }

  async get(traceId: string): Promise<TraceRecord | null> {
    const row = await this.db
      .prepare("SELECT * FROM traces WHERE request_id = ?")
      .bind(traceId)
      .first<TraceRow>();
    return row ? rowToTrace(row) : null;
  }

  async list(limit = 50): Promise<TraceRecord[]> {
    const result = await this.db
      .prepare("SELECT * FROM traces ORDER BY created_at DESC LIMIT ?")
      .bind(limit)
      .all<TraceRow>();
    return (result.results ?? []).map(rowToTrace);
  }

  async clear(): Promise<void> {
    await this.db.prepare("DELETE FROM traces").run();
  }
}

function rowToTrace(row: TraceRow): TraceRecord {
  const trace: TraceRecord = {
    trace_id: row.request_id,
    started_at: Date.parse(row.created_at) || 0,
    client_protocol: "gateway",
    status: row.status >= 200 && row.status < 400 ? "success" : "error",
    latency_ms: row.latency_ms,
    usage: {
      inputTokens: row.prompt_tokens,
      outputTokens: row.completion_tokens,
      totalTokens: row.total_tokens,
    },
    steps: parseJsonColumn(row.steps_json, []),
    request: parseJsonColumn(row.request_json, {}),
    response: parseJsonColumn(row.response_json, {}),
  };
  if (row.alias) trace.incoming_model = row.alias;
  if (row.provider_id != null) trace.upstream_provider_id = String(row.provider_id);
  if (row.target_model) trace.upstream_model = row.target_model;
  return trace;
}
